### PR TITLE
[fbx] Port FBX module from 3.2 branch

### PR DESCRIPTION
**Pull request by RevoluPowered**
Does what it says in the tin. Work in progress

Porting to 4.0 was delayed on purpose so we could land the many FBX fixes in the 3.2 branch so that further QC went on.

Supersedes #41985

Commit includes SHA so we know what branch has what version of the FBX module.

### CODE REVIEW

**Comment by RevoluPowered** [comment]
I took the opportunity to rewrite the FBX material handling like I promised for 4.0.

We are right now looking for a sensible default for our transparency properties. #44579 

I'll re-add the standard properties tomorrow and for 'transparency' most likely will just set alpha, unless we can make the same guarantee's as it had when it was exported from Blender, Maya or Max, which I currently isn't happening.

For emission, and various other things I can add that back in tomorrow it's fairly simple to do.

**Comment by RevoluPowered** [comment]
Legacy non PBR assets will be dealt with in later PR we are discussing some legacy components on a lot of assets, hang tight on replies someone is gonna write up some additional notes on our findings.

https://github.com/godotengine/godot-proposals/issues/2010

File: servers/rendering/renderer_storage.h [hunk 1]
```diff
@@ -71,7 +71,9 @@ class RendererStorage {
 			for (Set<InstanceDependency *>::Element *E = dependencies.front(); E; E = E->next()) {
 				InstanceDependency *dep = E->get();
 				Map<InstanceBaseDependency *, uint32_t>::Element *F = dep->instances.find(this);
-				ERR_CONTINUE(!F);
```
Comment (RevoluPowered) [inline servers/rendering/renderer_storage.h:74]: this should be made a separate PR

Comment (RevoluPowered) [reply servers/rendering/renderer_storage.h:74]: fixed.

File: modules/fbx/fbx_parser/LICENSE [hunk 2]
```diff
@@ -0,0 +1,76 @@
...
+on the use of their work. For any of these models, see
+<model-name>.source.txt for more legal information. Contact us if you
+are a copyright holder and believe that we credited you improperly or
+if you don't want your files to appear in the repository.
+
+
+******************************************************************************
+
+Poly2Tri Copyright (c) 2009-2010, Poly2Tri Contributors
+http://code.google.com/p/poly2tri/
```
Comment (akien-mga) [inline modules/fbx/fbx_parser/LICENSE:51]: I know it's the original Assimp licence, but these two sections don't apply to the code we're including so they can be removed.

I'd suggest adding a comment at the top of the file to indicate that files in the `fbx_parser` folder are derived from Assimp, and have been modified to match coding conventions and needs of the Godot project.

File: modules/fbx/readme.md [hunk 3]
```diff
@@ -0,0 +1,196 @@
+# Open Source FBX Specification for the Importer
```
Comment (akien-mga) [inline modules/fbx/readme.md:1]: I'd rename `readme.md` to `README.md`, which is more common.

Comment (RevoluPowered) [reply modules/fbx/readme.md]: done :)

File: modules/fbx/fbx_parser/LICENSE (hunk 2 above)
Comment (RevoluPowered) [reply modules/fbx/fbx_parser/LICENSE]: done both :)

**Review by akien-mga — APPROVED** [review]

**Comment by akien-mga** [comment]
Thanks!
